### PR TITLE
Fix data race on query_factories_info

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix data race on collecting factories info for system.query_log.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 9b8b1c50fa362c115174ba83e011c7efb7537b91 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59049/9b8b1c50fa362c115174ba83e011c7efb7537b91/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59049/9b8b1c50fa362c115174ba83e011c7efb7537b91/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Interpreters/Context.cpp [hunk 1]
```diff
@@ -1644,6 +1644,12 @@ void Context::addQueryAccessInfo(const QualifiedProjectionName & qualified_proje
         "{}.{}", qualified_projection_name.storage_id.getFullTableName(), backQuoteIfNeed(qualified_projection_name.projection_name)));
 }
 
+Context::QueryFactoriesInfo Context::getQueryFactoriesInfo() const
+{
+    std::lock_guard lock(query_factories_info.mutex);
```
Comment (devcrafter) [inline src/Interpreters/Context.cpp:1649]: 1It seems this lock is not necessary. The lock is taken in copy constructor on the right-hand object. See https://github.com/ClickHouse/ClickHouse/blob/9adab57771c75f691d4dec73fd3abc88d516206f/src/Interpreters/Context.h#L362 

Also, it seems assignment operator and swap() method are not in use - https://github.com/ClickHouse/ClickHouse/blob/9adab57771c75f691d4dec73fd3abc88d516206f/src/Interpreters/Context.h#L376-L393 . These methods don't take the lock. So, we either fix them, or better delete them.

**Review by devcrafter — APPROVED** [review]
LGTM with comment

**Comment by kssenii** [comment]
> AST fuzzer (tsan) — Received signal 11

https://github.com/ClickHouse/ClickHouse/issues/59107

> Stateless tests (release, analyzer) — Some queries hung, fail: 1

`02901_parallel_replicas_rollup` cc @Algunenano (as I see this test was added by you)

> Upgrade check (debug) 

`New settings are not reflected in settings changes history` is about `max_estimated_execution_time`, seems unrelated to this PR.

> Performance Comparison [1/4] — 2 errors, 1 too long, 1 faster, 1 unstable

Seems unrelated.

**Comment by Algunenano** [comment]
> 02901_parallel_replicas_rollup cc @Algunenano (as I see this test was added by you)

Already reverted the changes that introduced flakiness from master

**Pull request merged**